promql/engine.go

```diff
@@ -705,11 +734,12 @@ func (ev *evaluator) rangeEval(f func([]Value, *EvalNodeHelper) Vector, exprs ..
 			vectors[i] = vectors[i][:0]
 			for si, series := range matrixes[i] {
 				for _, point := range series.Points {
-					if point.T == ts {
+					if point.T == ts && inputSamples < ev.maxSamples {
 						vectors[i] = append(vectors[i], Sample{Metric: series.Metric, Point: point})
 						// Move input vectors forward so we don't have to re-scan the same
 						// past points at the next step.
 						matrixes[i][si].Points = series.Points[1:]
+						inputSamples++
```

Comment:
That's not right, the input samples are the entire of the input matrix - not jut the bits we've used so far.

Reply:
I'm not sure what you mean, because we already have `matrixes` in memory? We're only appending the samples we end up using to `vectors`, which is then used as `args` to `f`,

Reply:
Yes, they're already in memory.

Reply:
So then if the length of all the `Matrix` in `matrixes` is `> maxSamples` should we not just fail right away? Or only if `inputSamples + len(matrixes[i]) > maxSamples`?

Reply:
We should have already failed when one of them was being calculated.

Reply:
Right, so if `inputSamples + len matrixes[i] > maxSamples` we should fail, that should happen on line 707

Reply:
Yeah around, there. It'd be inside the eval.

Reply:
Well we're already tracking if each call to eval would load more than `maxSamples`, but not if the result of all the calls in rangeEval to eval would.

Reply:
So it seems if we track the # of samples that way, based on the outputs of the eval's, we don't need to track samples where I am here at all. That, in combination with not tracking samples when adding to the output series (https://github.com/prometheus/prometheus/pull/4513/commits/29e1469c81a1476a51c7b5f7ede128bb61ac7005), are we missing any cases? The engine tests I added would all still pass.